toonz/sources/tnztools/assistants/assistantellipse.cpp

```diff
@@ -0,0 +1,473 @@
+    if (r2 <= TConsts::epsilon) return TAffine::zero();
+    return TAffine::translation(m_center.position) *
+           TAffine::rotation(atan(da)) * TAffine::scale(r1, r2);
+  }
+
+  void getGuidelines(const TPointD &position, const TAffine &toTool,
+                     TGuidelineList &outGuidelines) const override {
+    bool restrictA = getRestrictA();
+    bool restrictB = getRestrictB();
+    bool repeat    = getRepeat();
+
+    TAffine matrix = calcEllipseMatrix();
+    if (matrix.isZero()) return;
+    if (!restrictA && restrictB) {
+      std::swap(matrix.rowX(), matrix.rowY());
+      std::swap(restrictA, restrictB);
+    }
+
+    matrix            = toTool * matrix;
+    TAffine matrixInv = matrix.inv();
+
+    if (restrictA && restrictB) {
+      // ellipse
+      outGuidelines.push_back(TGuidelineP(new TGuidelineEllipse(
+          getEnabled(), getMagnetism(), matrix, matrixInv)));
+    } else if (!restrictA && !restrictB) {
+      // scaled ellipse
+      TPointD p = matrixInv * position;
+      double l  = norm(p);
+      outGuidelines.push_back(TGuidelineP(new TGuidelineEllipse(
+          getEnabled(), getMagnetism(), matrix * TAffine::scale(l))));
+    } else {  // restrictA
+      TPointD p = matrixInv * position;
+      if (repeat) {
+        double ox = round(0.5 * p.x) * 2.0;
+        p.x -= ox;
+        matrix *= TAffine::translation(ox, 0.0);
+      }
+
+      // scale by Y
+      if (p.x <= TConsts::epsilon - 1.0) {
+        // line x = -1
+        outGuidelines.push_back(TGuidelineP(new TGuidelineInfiniteLine(
+            getEnabled(), getMagnetism(), matrix * TPointD(-1.0, 0.0),
+            matrix * TPointD(-1.0, 1.0))));
+      } else if (p.x >= 1.0 - TConsts::epsilon) {
+        // line x = 1
+        outGuidelines.push_back(TGuidelineP(new TGuidelineInfiniteLine(
+            getEnabled(), getMagnetism(), matrix * TPointD(1.0, 0.0),
+            matrix * TPointD(1.0, 1.0))));
+      } else {
+        // ellipse scaled by Y
+        double k = fabs(p.y / sqrt(1.0 - p.x * p.x));
+        outGuidelines.push_back(TGuidelineP(new TGuidelineEllipse(
+            getEnabled(), getMagnetism(), matrix * TAffine::scale(1.0, k))));
+      }
+    }
+  }
+
+private:
+  void drawEllipseRanges(const TAngleRangeSet &ranges,
+                         const TAffine &ellipseMatrix,
+                         const TAffine &screenMatrixInv, double pixelSize,
+                         double alpha) const {
+    assert(ranges.check());
+    TAngleRangeSet actualRanges(ranges);
+    const TRectD oneBox(-1.0, -1.0, 1.0, 1.0);
+    if (!TGuidelineEllipse::truncateEllipse(
+            actualRanges, ellipseMatrix.inv() * screenMatrixInv, oneBox))
+      return;
+    assert(actualRanges.check());
+
+    int segments =
+        TGuidelineEllipse::calcSegmentsCount(ellipseMatrix, pixelSize);
+    double da = M_2PI / segments;
+    double s  = sin(da);
+    double c  = cos(da);
+
+    for (TAngleRangeSet::Iterator i(actualRanges); i; ++i) {
```

Comment:
I am still trying to figure out the macOS crashes occuring in this procedure.  Problem is debugging through Xcode doesn't cause the issue.  I have to compile and run from command line to cause the crash but it's a bit painful to get any type of debugging information written somewhere I can see what it's doing.

I think part of the issue here is how you are iterating through this structure.  It assume iterator action `++i` will eventually come back as 0, but that isn't always a guarantee and there are no other checks to make sure you are not exceeding the bounds of the structure/memory, hence the crash.

Case in point: if i only allow it to iterate 1x through this for loop, it does not crash here.  I will crash in `guidelineellipse.cpp` -> `draw()`, where there is a similar for loop through `TAngleRangeSet`.  If I don't go through the loop in guidelineelipse, it doesn't crash. This might be because at creation time there are no guidelines to draw.

Reply:
So I can create an ellipse high up on the viewer edge. When I slowly drag it down, it will draw the ellipse.  But the second the full ellipse is drawn (actual ellipse or even the ellipse guideline) it crashes. It doesn't crash while running through xcode only happens on command line.  Not sure why that is but I'm using that to get some information shown here:

![image](https://user-images.githubusercontent.com/19245851/71545650-e6d98800-295b-11ea-899a-49e5b7628c84.png)

The image shows the values of the various values in this loop as I am dragging the ellipse from the top of the viewer down.  For testing purposes I commented out drawing the ellipse guideline.  The moment the full ellipse can be displayed in its entirety is when it crashes.  This seems to happens when `cnt` jumps from `90` to `94`.  Notice the `r.y` value is `-1.22465e-16`. 

The values are the same on Windows, but it doesn't crash there.  Somehow macOS isn't handling something like we would expect it to.

Not much more I can do with this.  No idea how to fix it.  If there is something you want me to try, please let me know.

In the next OTX release, I will try to disable the ellipse assistant for macOS builds until this is fixed.